hashexpr/src/lib.rs

```diff
@@ -172,7 +180,7 @@ impl Expr {
       let (i_type, type_code) = take(type_len)(i)?;
       let (i, data_len) = take(data_len_len)(i_type)?;
       let data_bitlen =
-        data_len.iter().fold(0, |acc, &x| (acc * 256) + x as u64);
+        data_len.iter().fold(0, |acc, &x| (acc * 256) + u64::from(x));
```

Comment:
`as` is inherently not unsafe, but unsound. There are lots of errors that can result including overflow, truncation, etc. Using `u64::from` is guaranteed to be sound, which is not true for other cases. Threrefore, where there is a from instance, it is better to use that instance to show that it is sound.

Reply:
So, I'm still not sure why `u64::from` would be guaranteed to be sound here but `as` is not. We have `x: u8`, and afaiu whenever we convert a small unsigned int to large one `as` just zero-extends.: https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions, which can't overflow or truncate.

I agree that for lots of cases `as` has a ton of implicit behavior, so if the intent here is to establish a general practice for using `From` instead of `as` where possible, then I think that's okay.

Reply:
I think you might know more than me on this topic, so I'll defer. My understanding was that Rust only provides a `From` instance when the conversion is sound. In this case the conversion is sound, so it makes sense to use `from` instead of `as` since an instance exists. If this practice is followed `as` becomes an indicator that there is possibly unsound behavior there, but the places where `from` is used, bugs should not be expected.

Reply:
That makes sense, happy to accept `from` as the safer convention